Arm backend: Add initial module tests for Stable Diffusion 3.5 Medium

YufengShi-dudu opened the pull request:
- Add module tests for two text encoders: CLIP and T5
- Add module tests for VAE autoencoder
- Add module tests for SD3Transformer2DModel
- Add flag to exmaples/arm/setup.sh for installing Stable Diffusion dependencies
- Handle int64 inputs to aten.slice_copy.Tensor using pass InsertCastForOpsWithInt64InputPass

Change-Id: I4389e87749cfb4e40f837cc6bfa8a59a73fb3a3a


cc @digantdesai @freddan80 @per @zingo @oscarandersson8218

YufengShi-dudu committed:
Arm backend: Add initial module tests for Stable Diffusion 3.5 Medium

- Add module tests for two text encoders: CLIP and T5
- Add module tests for VAE autoencoder
- Add module tests for SD3Transformer2DModel
- Add flag to exmaples/arm/setup.sh for installing Stable Diffusion
  dependencies
- Handle int64 inputs to aten.slice_copy.Tensor using pass
  InsertCastForOpsWithInt64InputPass

Change-Id: I4389e87749cfb4e40f837cc6bfa8a59a73fb3a3a
Signed-off-by: Yufeng Shi <yufeng.shi@arm.com>

YufengShi-dudu commented:
Hi @digantdesai,

We added some test configs to stable_diffusion_module_test_configs.py. They're mostly copy-pasted from the diffusers library to initialize different modules for testing.

We've included license notice and copyright attribution, and also links to the original source files for reference,

Please let us know if you have any concerns.

digantdesai commented on backends/arm/scripts/install_models_for_test.sh:
```diff
@@ -0,0 +1,10 @@
+#!/usr/bin/env bash
```
What about updating this file? - https://github.com/pytorch/executorch/blob/de0554d86c7f2dc8724e85fab6c809cd7d7c7bb5/requirements-examples.txt

Also we have this :(
https://github.com/pytorch/executorch/blob/de0554d86c7f2dc8724e85fab6c809cd7d7c7bb5/examples/qualcomm/qaihub_scripts/stable_diffusion/install_requirements.sh#L2

digantdesai replied:
The TODO says it all. Thanks.

digantdesai commented on backends/arm/_passes/insert_int64_input_cast_pass.py:
```diff
@@ -60,35 +66,59 @@ def call(self, graph_module):
 
-            if valid_for_insert:
+                if valid_for_insert:
+                    to_copy_op = self.get_decomposition(node.target)
+                    with graph.inserting_before(node):
+                        cast_before = create_node(
+                            graph,
+                            to_copy_op,
+                            args=(indices,),
+                            kwargs={
+                                "dtype": torch.int32,
+                                "memory_format": torch.preserve_format,
+                            },
+                        )
+                        node.replace_input_with(indices, cast_before)
+
+                    modified_graph = True
```
deduplicate with elif below?

digantdesai commented on backends/arm/README.md:
```diff
@@ -189,7 +197,14 @@ Configuration of the EthosUBackend export flow is controlled by CompileSpec info
-In order to lower this to TOSA and int64->int32 cast need to be injected. This pass need to run very early in the lowering process and can be passed in to the to_edge_transform_and_lower() function call as an optional parameter. See example in: backends/arm/test/models/test_llama.py.
-By doing this aten.embedding will be decomposed into to aten.index_select which can handle int32 indices.
-Note that this additional step is only needed for pure float models. With quantization this is automatically handled during annotation before the export stage.
+The current TOSA version does not support int64. However, int64 is commonly used in many models. In order to lower the operators with int64 inputs and/or outputs to TOSA, a few passes have been developed to handle the int64-related issues. The main idea behind these passes is to replace the uses of int64 with int32 where feasible.
```
this is also the case with xnnpack and other backends, perhaps we should share this pass.

YufengShi-dudu replied:
Issue created: https://github.com/pytorch/executorch/issues/12957

digantdesai commented on backends/arm/README.md:
```diff
@@ -189,7 +197,14 @@ Configuration of the EthosUBackend export flow is controlled by CompileSpec info
-In order to lower this to TOSA and int64->int32 cast need to be injected. This pass need to run very early in the lowering process and can be passed in to the to_edge_transform_and_lower() function call as an optional parameter. See example in: backends/arm/test/models/test_llama.py.
-By doing this aten.embedding will be decomposed into to aten.index_select which can handle int32 indices.
-Note that this additional step is only needed for pure float models. With quantization this is automatically handled during annotation before the export stage.
+The current TOSA version does not support int64. However, int64 is commonly used in many models. In order to lower the operators with int64 inputs and/or outputs to TOSA, a few passes have been developed to handle the int64-related issues. The main idea behind these passes is to replace the uses of int64 with int32 where feasible.
```
Love this documentation BTW!

digantdesai commented on examples/arm/setup.sh:
```diff
@@ -179,6 +179,11 @@ function check_options() {
                     exit 1
                 fi
                 ;;
+            --setup-test-dependency)
```
let's see if we can overload et's requirements-examples.txt
if you want we can do in a follow up PR

YufengShi-dudu replied:
Agree to have a separate requirements.txt at the end.

digantdesai reviewed (approved):
Great to see this test working :)

YufengShi-dudu committed:
Arm backend: Remove duplicated code when inserting int32 cast node

- Extract the logic of inserting int32 cast node into a helper
  function in pass InsertCastForOpsWithInt64InputPass

Change-Id: I748d53921981d43c3a913c4e7a15f3daf33c836e
Signed-off-by: Yufeng Shi <yufeng.shi@arm.com>

YufengShi-dudu committed:
Merge branch 'main' into stable-diffusion-module-tests

Sebastian-Larsson reviewed (approved):
Unrelated CI failures. Approved.
